developer-guide.md

```diff
@@ -0,0 +1,353 @@
+      <incoming-image(site-header) />
+    </image-wrapper(site-header)>
+  </container(site-header)>
+</special-top-layer>
+```
+
+Although, it's possible to have one of these structures that doesn't have an incoming/outgoing image. For example, if the element was only there before the DOM change, or only there after the DOM change.
+
+In a similar way that we were using `::page-transition-outgoing-image(root)` to target the outgoing image of the root, we can use `::page-transition-outgoing-image(site-header)` to target the outgoing image of the site-header, and animate it separately.
+
+The default animation animates the `width` and `height` of the `::page-transition-container` from its before size to its after size, while animating a `transform` to move it from its before position to its after position. If the header is unchanged, it'll appear fixed in place while the rest of the content (the root) slides as we specified earlier.
+
+### Flat transition structures
+
+Each captured element, along with the root, will be a direct child of the special top layer. This means you can move elements between containers, even if they have `overflow: hidden` or some other form of clipping in the real DOM.
+
+We plan to add a feature to allow one transition container to be nested within another, but that isn't currently implemented.
+
+## Animation synchronization gotcha
+
+Right now, there's a delay between Chrome creating the outgoing image element, and the incoming image element. That means the styles for the outgoing image element will apply sooner, which in turn means its animation will start sooner, and apply out of sync with the incoming image element.
+
+This is considered a bug (well, a design error), but you can work around it:
+
+```js
+async function spaNavigate(data) {
+  // Fallback
+  if (!document.createDocumentTransition) {
+    await updateDOMForPage(data);
+    return;
+  }
+
+  // With a transition
+  const transition = document.createDocumentTransition();
+
+  // Add a temporary class:
+  document.documentElement.classList.add("transition-warming-up");
+
+  await transition.start(async () => {
+    await updateDOMForPage(data);
+
+    // Now remove it:
+    document.documentElement.classList.remove("transition-warming-up");
+  });
+}
+```
+
+Then in the CSS:
+
+```css
+.transition-warming-up::page-transition-container(*),
+.transition-warming-up::page-transition-incoming-image(*),
+.transition-warming-up::page-transition-outgoing-image(*) {
+  animation-play-state: paused !important;
+}
+```
+
+This pauses the animations until both the outgoing and incoming content is ready.
+
+## Using `object-fit` and `object-position`
+
+The default animation animates the `width` and `height` of the `::page-transition-container`, which is generally frowned upon in web performance circles, as it runs layout per frame. However, for page transitions, we plan to optimize it so it isn't an issue (it isn't optimized yet).
```

Comment:
Not sure if we need to mention: "which is generally frowned upon in web performance circles, as it runs layout per frame. However, for page transitions, we plan to optimize it so it isn't an issue (it isn't optimized yet)."

I thought its frowned upon coz running layout on complex trees is slow but in this case it's about running on a thread with possibly lots of contention. We could clarify but it wasn't clear if this detail is needed here.

Reply:
I think if we don't mention it, some smartarse will be on Twitter saying "hahaha Google so dumb, they don't even follow their own performance advice"